### DISCUSSION TITLE
Should filtering options be defined per Postgres or Mathesar type?

### DISCUSSION THREAD

**Discussion by dmos62** [mathesar-foundation/mathesar#843] [Maintainer discussions]
Part of the filtering functionality is to expose in the REST API the fitlering options available to each type. At the moment, our REST API only exposes metadata on the Mathesar types. However, @kgodey [indicated](https://github.com/centerofci/mathesar/issues/385#issuecomment-972932443) that filtering options should be defined per Postgres type, which seems to contradict that architecture.

@mathemancer provided some input in chat, saying that filtering on Mathesar types should be a good abstraction:

> I can't think of a place where I'd want a filter on a DB type that wouldn't be appropriate for all other DB types under a given Mathesar type.  In fact, it seems that if we needed different filters there, it might be a signal that the Mathesar type should actually be multiple types.

I tend to agree, though I'm wondering if there's more to this.

**Comment by pavish** [comment]
I agree with @mathemancer's point as well. If the db types requires different filtering options, they should probably belong to a different Mathesar type.

I am wondering if there are any possible cases where grouping differs by db type for types within the same mathesar type. It would be better if we can bring in both filtering and grouping options within Mathesar types.

**Comment by kgodey** [comment]
Here's my reasoning for filtering and grouping being on DB types:
- Some Mathesar types that accept different filters e.g. "Date & Time" type which contains the `TIMESTAMP`, `TIME`, and `DATE` types. The filter parameters for each are not the same. e.g. you cannot pass a date to a `TIME` type or vice versa.
- We're building the DB module and API to support non-Mathesar UIs. People should be able to use our APIs without buying into our abstractions. If someone wants to build a UI that ignores the concept of Mathesar types entirely, they should be able to do so. We can't support that if filtering and grouping depends on Mathesar types.

**Reply by dmos62** [reply]
Those are some interesting points. Your first point about Date & Time Mathesar type is what others mentioned as a sign that this Mathesar type should be split into multiple.

I feel like I've not yet fully wrapped my head around your second point. This is not the first time it's come up. I might see a way that this is doable without tangling up the architecture: we primarily define the filtering options on Mathesar types, but then (since every Postgres type is associated with only one Mathesar type), we can derive a secondary definition of what are the options for a given Postgres type. This has the advantage of relying on a definition of a Mathesar type that involves homogenous filtering (you can use the same filtering operations on all Postgres types encapsulated by this Mathesar type).

**Reply by kgodey** [reply]
@dmos62 We could split Date & Time into "Date", "Time" and "Date & Time" types, but my concerns with that are:
- We'd need to redo the design work for it (see the [Date & Time prototype](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4260%3A37440&node-id=4270%3A41050&viewport=324%2C48%2C0.29&scaling=contain&starting-point-node-id=4270%3A41050&show-proto-sidebar=1)).
- Eventually, we plan to support a _lot_ of different data types. This could become overwhelming to the user if we don't group them in some way. I think "Date & Time" is a natural grouping.

> we can derive a secondary definition of what are the options for a given Postgres type. 

This seems fine to me.

**Reply by dmos62** [reply]
If we want the Mathesar type set to scale well then it's even more important to have a strong definition for what a Mathesar type is. The more different operations (like filtering) we can perform on a Mathesar type (without falling back to examining the underlying Postgres type) the stronger its definition. That's good for the user too, since it makes the interface simpler.

**Reply by kgodey** [reply]
Makes sense. We also have an ongoing discussion about potentially extending the Mathesar type system to View columns in #838. 

I think a good next step might be for me to create a "Concepts" page for data types similar to [the Views page](https://wiki.mathesar.org/en/product/concepts/views) and open that up for feedback so we can all be on the same page. That should help us both define the types better and give us a definition to rely on for further work.

**Reply by mathemancer** [reply]
I think the implementation of filtering should be in the DB layer, using SQL functions called from python.  I wasn't really thinking of filters as being applied to Mathesar types implementation-wise.  From this perspective, then the Mathesar type would provide a grouping of DB types where the operations would be consistent within a group of types.  I see @kgodey 's point about this not quite working, though.  

I wonder if we should eventually move towards the "Mathesar type" being "type group" instead?  That way, the user could pick a type group, and get some automatic defaults (e.g., the database type and associated filters, sorting functions, aggregations, etc.).  If they want, they could choose a different type in the group.  It might also be possible (someday) for them to configure what type is the default for a given group.  They could also configure some group-wide options in the DB (e.g., sorting).

@dmos62 if you create that page, please toss a link somewhere I might notice it.  I agree that we should get these things hashed out pretty precisely at this juncture.

**Reply by dmos62** [reply]
@mathemancer that's quite exactly how I've been thinking about Mathesar types. I agree about filtering happening on Postgres. For the foreseeable future it should all be based on our fork of `sqlalchemy-filters`.

@kgodey I think some well targeted GitHub discussion threads would be a good compliment to starting that wiki page for types. I'll start one if a good prompt comes to mind.

**Reply by kgodey** [reply]
@dmos62 Sounds good.
